### ISSUE TITLE
$global improperly defined in webextension

### ISSUE THREAD

**Issue by mwalcott3** [gopherjs/gopherjs#724]
**Summary:** 

- Using "this" is a better option than "window" when running in a browser extension (Using "window" breaks the script in my case)
- A new test case needs to be added to see if the script is running in a browser extension

**Explanation:**
Note: This was a Firefox extension I haven't tested the behavior in chrome extension yet. 

When running a script in a webextention I was getting the error "TypeError: $global.String.fromCharCode is not a function" So I suspected that global was improperly set.

Here is the code used to define $global in a file generated by gopherjs:
```
var $global, $module;
if (typeof window !== "undefined") { /* web page */
  $global = window;
} else if (typeof self !== "undefined") { /* web worker */
  $global = self;
} else if (typeof global !== "undefined") { /* Node.js */
  $global = global;
  $global.require = require;
} else { /* others (e.g. Nashorn) */
  $global = this;
}

if ($global === undefined || $global.Array === undefined) {
  throw new Error("no global object found");
}
```

Here is a test I ran (Without the rest of the generated code):
```
var $global
if (typeof window !== "undefined") { /* web page */
	$global = window;
	console.log($global);
        //$global.String exists but $global.String.fromCharCode does not
	console.log($global.String.fromCharCode);
} else if (typeof self !== "undefined") { /* web worker */
  $global = self;
} else if (typeof global !== "undefined") { /* Node.js */
  $global = global;
  $global.require = require;
} else { /* others (e.g. Nashorn) */
  $global = this;
}
$global = this;
console.log($global);
console.log($global.String.fromCharCode);
```

This printed the following:
```
Window https://trakt.tv/movies/deadpool-2016
undefined
Sandbox { browser: {…}, chrome: {…}, window: Window, document: HTMLDocument https://trakt.tv/movies/deadpool-2016, location: Location https://trakt.tv/movies/deadpool-2016, top: Window, self: Window, name: "clicktoresponse:1512625427187:http://www.imdb.com/", … }
function ()
```
The problem is that while "window" exists in an extension it is somewhat different than when running normally. So it has enough attached to pass the current test ("$global.Array" exists. Hell, even $global.String exists ) but it doesn't have some other functions ("$global.String.fromCharCode" for instance). 

Manually editing the code to set "$global = this" fixed my extension.

**Comment by shurcooL** [comment]
This unfortunately sounds messy.

I wonder if there's a reasonable and safe way we can detect this, without compromising the behavior on other platforms.

Perhaps by checking if `typeof window.String.fromCharCode === "undefined"` and `typeof this.String.fromCharCode !== "undefined"`?

**Comment by mwalcott3** [comment]
Yeah. Making sure any change doesn't affect any other platforms should be the first priority. The fact that noone brought this up before suggests it isn't a problem for most current users of gopherjs.

Running a check like `window && window.String && window.String.fromCharCode` for each possible $global value should be fine. However, it should be tested on each platform first to verify it doesn't muck something up. 

Im just playing around with gopherjs so manually changing the generated file works for me.
